sdks/python/apache_beam/transforms/trigger.py

```diff
@@ -1038,14 +1098,22 @@ def process_timer(self, window_id, unused_name, time_domain, timestamp,
     window = state.get_window(window_id)
     if state.get_state(window, self.TOMBSTONE):
       return
-    if time_domain == TimeDomain.WATERMARK:
-      if not self.is_merging or window in state.known_windows():
-        context = state.at(window)
+
+    if not self.is_merging or window in state.known_windows():
+      context = state.at(window, self.clock)
+      if time_domain == TimeDomain.WATERMARK:
         if self.trigger_fn.should_fire(timestamp, window, context):
           finished = self.trigger_fn.on_fire(timestamp, window, context)
           yield self._output(window, finished, state)
-    else:
-      raise Exception('Unexpected time domain: %s' % time_domain)
+      elif time_domain == TimeDomain.REAL_TIME:
+        # TODO(mariagh): Move this evaluation to class AfterProcessingTime
```

Comment:
Please do this TODO now rather than leaving trigger-specific logic here. (In fact, it seems you should be able to use exactly the same code rather than branching on the time domain as the clock is part of context.)

Reply:
Note that we would need to add a `time_domain` argument to `should_fire` so that the trigger checks the correct time domain before firing.

Reply:
Right, that is what I mentioned in my earlier comment--having a clock is not an issue, having time_domain is. I would need to have TriggerFn take an extra arg (since we make the call via the parent class), do we want to do that? It is not difficult, just tedious. Is there a better way?

Reply:
That seems like the best way given our constraints.

Reply:
There's no need to pass the time domain. The triggers set the timer; it knows what the time domain should be when it gets it back. If a TriggerFn is setting timers in multiple time domains, they should have different names.

Reply:
The issue is that `should_fire` is called by the `TriggerDriver` with the current watermark when elements are processed to determine if a bundle should fire.  We need to support both this case and the timer firing case here.

Reply:
Ah, yes, we need this distinction for earlier calls. 

We only have a handful of triggers, I'd still rather do that now than leave this as a TODO. Alternatively you could put a should_fire_with_time_domain(timestamp, window, context, time_domain) that delegates to should_fire(timestamp, window, context) in the TriggerFn base class with a TODO to unify this.